### PR TITLE
Implement DurabilityData

### PR DESCRIPTION
**Pull request by Rain336**
Implementation of `SpongeDurabilityData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableSpongeDurabilityData`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `UnbreakableValueProcessor`
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Implementation of the `ItemDurabilityValueProcessor`
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `FIELD_NAME` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

Test Plugin code:
```java
package net.rain.test;

import org.spongepowered.api.data.key.Keys;
import org.spongepowered.api.data.manipulator.mutable.item.DurabilityData;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.game.state.GamePostInitializationEvent;
import org.spongepowered.api.item.inventory.ItemStack;
import org.spongepowered.api.plugin.Plugin;
import org.spongepowered.api.text.Texts;
import org.spongepowered.api.text.format.TextColors;
import org.spongepowered.api.util.command.CommandResult;
import org.spongepowered.api.util.command.args.GenericArguments;
import org.spongepowered.api.util.command.spec.CommandSpec;

@Plugin(id = "CodeTester", name = "CodeTester", version = "1.0.0")
public class CodeTester {

    @Listener
    public void postInit(GamePostInitializationEvent event) {
        DurabilityData data = event.getGame().getRegistry().getManipulatorRegistry().getBuilder(DurabilityData.class).get().create()
                .set(Keys.ITEM_DURABILITY, 10).set(Keys.UNBREAKABLE, true);

        CommandSpec testData = CommandSpec.builder()
                .description(Texts.of(TextColors.GREEN, "Tests a DataManipulator"))
                .arguments(GenericArguments.integer(Texts.of("Type")))
                .executor(((src, args) -> {
                    int i = args.<Integer>getOne("Type").get();
                    if (src instanceof Player) {
                        Player player = (Player) src;
                        if (player.getItemInHand().isPresent()) {
                            ItemStack item = player.getItemInHand().get();
                            if (i == 1) {
                                if (item.supports(Keys.ITEM_DURABILITY)) {
                                    int durability = item.get(Keys.ITEM_DURABILITY).get();
                                    player.sendMessage(Texts.of("Durability: ", durability));
                                    item.offer(Keys.ITEM_DURABILITY, durability - 5);
                                }
                                return CommandResult.success();
                            }
                            if (i == 2) {
                                if (item.supports(Keys.UNBREAKABLE)) {
                                    boolean unbreakable = item.get(Keys.UNBREAKABLE).get();
                                    player.sendMessage(Texts.of("Unbreakable: ", unbreakable));
                                    item.offer(Keys.UNBREAKABLE, unbreakable ? false : true);
                                }
                                return CommandResult.success();
                            }
                            if (i == 3) {
                                if (item.supports(data.durability()) && item.supports(data.unbreakable())) {
                                    item.offer(data);
                                }
                                return CommandResult.success();
                            }
                            player.setItemInHand(item);
                        }
                    }
                    return CommandResult.empty();
                })).build();
        event.getGame().getCommandDispatcher().register(this, testData, "datatest");
    }
}
 ```

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        this.registerGetters();
+    }
+
+    public ImmutableSpongeDurabilityData() {
+        this(60, false);
+    }
+
+    @Override
+    protected void registerGetters() {
+        registerFieldGetter(Keys.ITEM_DURABILITY, () -> this.durability);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java]: Should be just a method reference like the others.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        registerKeyValue(Keys.ITEM_DURABILITY, this::durability);
+
+        registerFieldGetter(Keys.UNBREAKABLE, this.unbreakable::get);
+        registerKeyValue(Keys.UNBREAKABLE, this::unbreakable);
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> durability() {
+        return new ImmutableSpongeBoundedValue<Integer>(Keys.ITEM_DURABILITY, this.durability, 60, intComparator(), 0, Integer.MAX_VALUE);
```
Comment (ZephireNZ) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java]: Use a final class variable so you don't create this object each time.

Also, use the `SpongeValueBuilder` as well.

Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java]: store to final field locally, no need to keep returning new instances. Also, I recommend using the `SpongeValueBuilder.boundedBuilder()` to make a mutable and then call `asImmutable()`.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/item/SpongeDurabilityData.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        registerFieldSetter(Keys.ITEM_DURABILITY, this::setDurability);
+        registerKeyValue(Keys.ITEM_DURABILITY, this::durability);
+
+        registerFieldGetter(Keys.UNBREAKABLE, () -> this.unbreakable);
+        registerFieldSetter(Keys.UNBREAKABLE, value -> this.unbreakable = value);
+        registerKeyValue(Keys.UNBREAKABLE, this::unbreakable);
+    }
+
+    public void setDurability(int durability) {
+        if (durability >= 0) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/item/SpongeDurabilityData.java]: replace with `checkArgument(durability >= 0, "Durability cannot be less than zero!");`

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/item/SpongeDurabilityData.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            throw new IllegalArgumentException("Invalid value for durability");
+        }
+    }
+
+    @Override
+    public MutableBoundedValue<Integer> durability() {
+        return SpongeValueBuilder.boundedBuilder(Keys.ITEM_DURABILITY)
+                .minimum(0)
+                .maximum(Integer.MAX_VALUE)
+                .defaultValue(60)
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/item/SpongeDurabilityData.java]: This probably won't be the default value for many items, I believe these should be dependent on the `ItemStack` being provided to the `DataProcessor`/`ValueProcessor`.

File: src/main/java/org/spongepowered/common/data/processor/data/item/DurabilityDataProcessor.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    public boolean set(ItemStack itemStack, Map<Key<?>, Object> keyValues) {
+        itemStack.setItemDamage(itemStack.getMaxDamage() - (int) keyValues.get(Keys.ITEM_DURABILITY));
+        final boolean unbreakable = (boolean) keyValues.get(Keys.UNBREAKABLE);
+        if (unbreakable) {
+            itemStack.setItemDamage(0);
+        }
+        if (!itemStack.hasTagCompound()) {
+            itemStack.setTagCompound(new NBTTagCompound());
+        }
+        itemStack.getTagCompound().setBoolean("Unbreakable", unbreakable);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/item/DurabilityDataProcessor.java]: Extract the string to `NbtDataUtil` please.

File: src/main/java/org/spongepowered/common/data/processor/value/item/ItemDurabilityValueProcessor.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    @Override
+    public boolean set(ItemStack container, Integer value) {
+        container.setItemDamage(container.getMaxDamage() - value);
+        return true;
+    }
+
+    @Override
+    public Optional<Integer> getVal(ItemStack container) {
+        return Optional.of(container.getMaxDamage() - container.getItemDamage());
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/item/ItemDurabilityValueProcessor.java]: You're not checking if the `ItemStack.isDamageable()`.

**Comment by ZephireNZ** [comment]
Should plugins be able to get/set the `DurablityData` of all items? I know that all items do technically have a durability, but the javadocs only talk about items which can be damaged (and that's really what this data is for, not setting the raw data value). Minecraft is moving away from most item damage anyway, so we should potentially keep that in mind to account for breaking changes in the future.

**Comment by gabizou** [comment]
> Should plugins be able to get/set the DurablityData of all items? I know that all items do technically have a durability, but the javadocs only talk about items which can be damaged (and that's really what this data is for, not setting the raw data value). Minecraft is moving away from most item damage anyway, so we should potentially keep that in mind to account for breaking changes in the future.

No, only `Item`s that can be actually damaged, not items that have subtypes, which is why `isDamageable()` is such an important check.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java (hunk 1 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java]: @gabizou What's the difference? AFAIK the compiler just converts lambdas to a method and replaces their usage with a call to the method.

Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java]: yes, but it's just cleaner.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/item/SpongeDurabilityData.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+import org.spongepowered.api.data.value.mutable.MutableBoundedValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.common.data.manipulator.immutable.item.ImmutableSpongeDurabilityData;
+import org.spongepowered.common.data.manipulator.mutable.common.AbstractData;
+import org.spongepowered.common.data.value.SpongeValueBuilder;
+
+public class SpongeDurabilityData extends AbstractData<DurabilityData, ImmutableDurabilityData> implements DurabilityData {
+
+    int durability;
+    boolean unbreakable;
```
Comment (Phroa) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/item/SpongeDurabilityData.java]: Should be private

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+public class ImmutableSpongeDurabilityData extends AbstractImmutableData<ImmutableDurabilityData, DurabilityData> implements ImmutableDurabilityData {
+
+    final ImmutableBoundedValue<Integer> durability;
+    final ImmutableValue<Boolean> unbreakable;
+
+    public ImmutableSpongeDurabilityData(int durability, boolean unbreakable) {
+        super(ImmutableDurabilityData.class);
+        checkArgument(durability >= 0);
+        this.durability = SpongeValueBuilder.boundedBuilder(Keys.ITEM_DURABILITY)
+                .minimum(0)
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/item/ImmutableSpongeDurabilityData.java]: missing a default value.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/item/SpongeDurabilityData.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    public SpongeDurabilityData() {
+        this(60, false);
+    }
+
+    public SpongeDurabilityData(int durability, boolean unbreakable) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/item/SpongeDurabilityData.java]: You should likely throw the default value in the constructor as well considering each itemstack has a custom durability setting.

File: src/main/java/org/spongepowered/common/data/processor/data/item/DurabilityDataProcessor.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+public class DurabilityDataProcessor extends AbstractItemDataProcessor<DurabilityData, ImmutableDurabilityData> {
+
+    public DurabilityDataProcessor() {
+        super(input -> input.getItem().isDamageable());
+    }
+
+    @Override
+    public boolean doesDataExist(ItemStack itemStack) {
+        return true;
```
Comment (phroa) [inline src/main/java/org/spongepowered/common/data/processor/data/item/DurabilityDataProcessor.java]: Do *all* itemstacks have some form of durability?

Comment (JBYoshi) [inline src/main/java/org/spongepowered/common/data/processor/data/item/DurabilityDataProcessor.java]: Should be `return itemStack.isDamageable();` (haven't tested, but there should be a similar method in `ItemStack`). Plugins shouldn't need to use damage values anymore unless the item is acually damageable.

**Pull request closed without merging**